dataservice/api/errors.py

```diff
@@ -1,45 +1,46 @@
+import re
 from dataservice.api.common.schemas import ErrorSchema
+from dataservice.extensions import db
 
-FOREIGN_KEY_ERROR_PREFIX = 'foreign key constraint failed'
-FOREIGN_KEY_ERROR_MSG_TEMPLATE = \
-    'Cannot {} {} without an existing {} entity'
 
-UNIQUE_CONSTRAINT_ERROR_PREFIX = 'unique constraint failed'
-UNIQUE_CONSTRAINT_ERROR_MSG_TEMPLATE = \
-    'Cannot {0} {1}, {2} already has a {1}'
-
-DEFAULT_INVALID_INPUT_TEMPLATE = 'Client error: invalid {} provided'
+FOREIGN_KEY_RE = re.compile(r'DETAIL:.*\((?P<kf_id>\w{8})\)' +
+                            ' is not present in table "(?P<table>\w+)"\.')
+UNIQUE_RE = re.compile(r'DETAIL:.*\((?P<entity>\w+)_id\)=' +
+                       '\((?P<kf_id>[A-Z0-9]{8})\) already exists\.')
 
 
 def http_error(e):
     """ Handles all HTTPExceptions """
     return ErrorSchema().jsonify(e), e.code
 
 
-def handle_integrity_error(**kwargs):
+def integrity_error(e):
     """
     Handle invalid client input error
 
     Determine error type based on error kwargs
     Return message based on kwargs
     """
-    # Default error message
-    message = DEFAULT_INVALID_INPUT_TEMPLATE.format(kwargs.get('entity'))
-
-    # Extract type of error from exception message
-    error_type = str(kwargs.get('exception').orig).strip().lower()
-    method = kwargs.get('method')
-    entity = kwargs.get('entity')
-    ref_entity = kwargs.get('ref_entity')
-
-    # Foreign key constraint failed
-    if FOREIGN_KEY_ERROR_PREFIX in error_type:
-        message = FOREIGN_KEY_ERROR_MSG_TEMPLATE.\
-            format(method, entity, ref_entity)
-
-    # Unique constraint failed
-    elif UNIQUE_CONSTRAINT_ERROR_PREFIX in error_type:
-        message = UNIQUE_CONSTRAINT_ERROR_MSG_TEMPLATE.\
-            format(method, entity, ref_entity)
-
-    return message
+    db.session.rollback()
+    error = [l for l in e.orig.pgerror.split('\n')
+             if l.startswith('DETAIL')][0]
+
+    message = 'error saving changes'
+
+    m = FOREIGN_KEY_RE.match(error)
+    # Tried to create entity, but another entity must be created first
+    if m and 'kf_id' in m.groups():
+        message = 'A {} must be created first'.format(m.group('table'))
+    # Tried to create an entity dependent on an entity that doesn't exist
+    elif m:
+        message = '{} "{}" does not exist'.format(m.group('table'),
+                                                  m.group('kf_id'))
+
+    m = UNIQUE_RE.match(error)
```

Comment:
Hmm the error message is a little confusing here. If someone tries to create a demographic for a participant that already has a demographic the error message is:  `participant "144WDN78" already exists`

Can we make it something like: `participant "144WDN78" already has demographic`.
You could get the primary entity (demographic) by parsing e.statement (i.e. INSERT INTO demographic, UPDATE demographic etc).